File: Runtime/Plugins/iOS/NuxieUnityBridge.swift
import Foundation

#if canImport(Nuxie)
import Nuxie
#endif

@_cdecl("NuxieUnity_Invoke")
public func NuxieUnity_Invoke(
  _ methodPointer: UnsafePointer<CChar>?,
  _ argumentsPointer: UnsafePointer<CChar>?,
  _ callbackObjectPointer: UnsafePointer<CChar>?,
  _ callbackMethodPointer: UnsafePointer<CChar>?
) -> UnsafeMutablePointer<CChar>? {
  let method = methodPointer.map(String.init(cString:)) ?? ""
  let arguments = argumentsPointer.map(String.init(cString:)) ?? "{}"
  let callbackObject = callbackObjectPointer.map(String.init(cString:)) ?? "__NuxieBridgeHost"
  let callbackMethod = callbackMethodPointer.map(String.init(cString:)) ?? "OnNuxieNativeEvent"

  return strdup(
    UnityNuxieBridge.shared.invoke(
      method: method,
      argumentsJSON: arguments,
      callbackObject: callbackObject,
      callbackMethod: callbackMethod
    )
  )
}

@_cdecl("NuxieUnity_FreeCString")
public func NuxieUnity_FreeCString(_ pointer: UnsafeMutablePointer<CChar>?) {
  free(pointer)
}

private final class UnityNuxieBridge: @unchecked Sendable {
  static let shared = UnityNuxieBridge()

  private let targetLock = NSLock()
  private var callbackObject = "__NuxieBridgeHost"
  private var callbackMethod = "OnNuxieNativeEvent"

  #if canImport(Nuxie)
  private lazy var purchaseDelegate = UnityPurchaseDelegateBridge(emit: emit)
  @MainActor private lazy var delegate = UnityDelegateBridge(emit: emit)
  #endif

  func invoke(
    method: String,
    argumentsJSON: String,
    callbackObject: String,
    callbackMethod: String
  ) -> String {
    setCallbackTarget(object: callbackObject, method: callbackMethod)
    guard let arguments = parseObject(argumentsJSON) else {
      return errorResponse(code: "INVALID_ARGUMENTS", message: "Arguments must be a JSON object.")
    }

    #if canImport(Nuxie)
    switch method {
    case "configure":
      guard let apiKey = arguments["apiKey"] as? String,
            !apiKey.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
      else {
        return errorResponse(code: "MISSING_API_KEY", message: "Nuxie API key is required.")
      }
      let options = UnsafeSendable(arguments["options"] as? [String: Any])
      let usePurchaseController = arguments["usingPurchaseController"] as? Bool ?? false
      return mapResult(blocking {
        try await MainActor.run {
          let configuration = self.makeConfiguration(
            apiKey: apiKey,
            options: options.value,
            usePurchaseController: usePurchaseController
          )
          NuxieSDK.shared.delegate = self.delegate
          do {
            try NuxieSDK.shared.setup(with: configuration)
            return true
          } catch {
            NuxieSDK.shared.delegate = nil
            throw error
          }
        }
      }) { _ in nil }

    case "shutdown":
      purchaseDelegate.cancelPending(reason: "sdk_shutdown")
      return mapResult(blocking {
        await NuxieSDK.shared.shutdown()
        await MainActor.run {
          NuxieSDK.shared.delegate = nil
        }
        return true
      }) { _ in nil }

    case "identify":
      guard let distinctId = arguments["distinctId"] as? String else {
        return errorResponse(code: "INVALID_ARGUMENTS", message: "identify requires distinctId.")
      }
      NuxieSDK.shared.identify(
        distinctId,
        userProperties: arguments["userProperties"] as? [String: Any],
        userPropertiesSetOnce: arguments["userPropertiesSetOnce"] as? [String: Any]
      )
      return okResponse(nil)

    case "reset":
      NuxieSDK.shared.reset(keepAnonymousId: arguments["keepAnonymousId"] as? Bool ?? false)
      return okResponse(nil)

    case "getDistinctId":
      return okResponse(NuxieSDK.shared.getDistinctId())

    case "getAnonymousId":
      return okResponse(NuxieSDK.shared.getAnonymousId())

    case "getIsIdentified":
      return okResponse(NuxieSDK.shared.isIdentified)

    case "trigger":
      guard let eventName = arguments["eventName"] as? String else {
        return errorResponse(code: "INVALID_ARGUMENTS", message: "trigger requires eventName.")
      }
      NuxieSDK.shared.trigger(
        eventName,
        properties: arguments["properties"] as? [String: Any]
      )
      return okResponse(nil)

    case "dismiss":
      return mapResult(blocking {
        await NuxieSDK.shared.dismiss()
        return true
      }) { _ in nil }

    case "setLocaleIdentifier":
      let localeIdentifier = arguments["localeIdentifier"] as? String
      return mapResult(blocking {
        try await NuxieSDK.shared.setLocaleIdentifier(localeIdentifier)
        return true
      }) { _ in nil }

    case "hasFeature":
      guard let featureId = arguments["featureId"] as? String else {
        return errorResponse(code: "INVALID_ARGUMENTS", message: "hasFeature requires featureId.")
      }
      let requiredBalance = number(arguments["requiredBalance"]) ?? 1
      let entityId = arguments["entityId"] as? String
      let useRemotePolicy = arguments["policy"] as? String == "remote"
      return mapResult(blocking {
        let access = try await NuxieSDK.shared.hasFeature(
          featureId,
          requiredBalance: requiredBalance,
          entityId: entityId,
          policy: useRemotePolicy ? .remote : .cacheFirst
        )
        return self.featureAccessDictionary(access)
      })

    case "useFeature":
      guard let featureId = arguments["featureId"] as? String else {
        return errorResponse(code: "INVALID_ARGUMENTS", message: "useFeature requires featureId.")
      }
      NuxieSDK.shared.useFeature(
        featureId,
        amount: number(arguments["amount"]) ?? 1,
        entityId: arguments["entityId"] as? String,
        metadata: arguments["metadata"] as? [String: Any]
      )
      return okResponse(nil)

    case "useFeatureAndWait":
      guard let featureId = arguments["featureId"] as? String else {
        return errorResponse(code: "INVALID_ARGUMENTS", message: "useFeatureAndWait requires featureId.")
      }
      let amount = number(arguments["amount"]) ?? 1
      let entityId = arguments["entityId"] as? String
      let setUsage = arguments["setUsage"] as? Bool ?? false
      let metadata = UnsafeSendable(arguments["metadata"] as? [String: Any])
      return mapResult(blocking {
        let result = try await NuxieSDK.shared.useFeatureAndWait(
          featureId,
          amount: amount,
          entityId: entityId,
          setUsage: setUsage,
          metadata: metadata.value
        )
        return self.featureUsageResultDictionary(result)
      })

    case "completePurchase":
      guard let requestId = arguments["requestId"] as? String,
            let payload = arguments["result"] as? [String: Any]
      else {
        return errorResponse(
          code: "INVALID_ARGUMENTS",
          message: "completePurchase requires requestId and result."
        )
      }
      purchaseDelegate.completePurchase(requestId: requestId, payload: payload)
      return okResponse(nil)

    case "completeRestore":
      guard let requestId = arguments["requestId"] as? String,
            let payload = arguments["result"] as? [String: Any]
      else {
        return errorResponse(
          code: "INVALID_ARGUMENTS",
          message: "completeRestore requires requestId and result."
        )
      }
      purchaseDelegate.completeRestore(requestId: requestId, payload: payload)
      return okResponse(nil)

    default:
      return errorResponse(code: "NATIVE_ERROR", message: "Unsupported method '\(method)'.")
    }
    #else
    return errorResponse(code: "NATIVE_UNAVAILABLE", message: "The Nuxie iOS SDK is not linked.")
    #endif
  }

  private func setCallbackTarget(object: String, method: String) {
    targetLock.withLock {
      callbackObject = object
      callbackMethod = method
    }
  }

  private func callbackTarget() -> (String, String) {
    targetLock.withLock { (callbackObject, callbackMethod) }
  }

  private func parseObject(_ raw: String) -> [String: Any]? {
    guard let data = raw.data(using: .utf8),
          let value = try? JSONSerialization.jsonObject(with: data)
    else {
      return nil
    }
    return value as? [String: Any]
  }

  private func number(_ value: Any?) -> Double? {
    (value as? NSNumber)?.doubleValue
  }

  #if canImport(Nuxie)
  @MainActor
  private func makeConfiguration(
    apiKey: String,
    options: [String: Any]?,
    usePurchaseController: Bool
  ) -> NuxieConfiguration {
    let configuration = NuxieConfiguration(apiKey: apiKey)

    if options?["environment"] as? String == "development" {
      configuration.environment = .development
    } else if options?["environment"] != nil {
      configuration.environment = .production
    }

    if let logLevel = options?["logLevel"] as? String {
      switch logLevel {
      case "verbose": configuration.logLevel = .verbose
      case "debug": configuration.logLevel = .debug
      case "info": configuration.logLevel = .info
      case "error": configuration.logLevel = .error
      case "none": configuration.logLevel = .none
      default: configuration.logLevel = .warning
      }
    }

    if let value = options?["enableConsoleLogging"] as? Bool {
      configuration.enableConsoleLogging = value
    }
    if let value = options?["redactSensitiveData"] as? Bool {
      configuration.redactSensitiveData = value
    }
    if options?.keys.contains("localeIdentifier") == true {
      configuration.localeIdentifier = options?["localeIdentifier"] as? String
    }
    if let value = options?["purchaseHandlingMode"] as? String {
      configuration.purchaseHandlingMode = value == "observer" ? .observer : .full
    }
    if let value = options?["testStoreEnabled"] as? Bool {
      configuration.testStoreEnabled = value
    }
    if usePurchaseController {
      configuration.purchaseDelegate = purchaseDelegate
    }

    return configuration
  }

  private func featureAccessDictionary(_ access: FeatureAccess) -> [String: Any] {
    [
      "allowed": access.allowed,
      "unlimited": access.unlimited,
      "balance": nullable(access.balance),
      "type": access.type.rawValue,
    ]
  }

  private func featureUsageResultDictionary(_ result: FeatureUsageResult) -> [String: Any] {
    let usage: Any
    if let value = result.usage {
      usage = [
        "current": value.current,
        "limit": nullable(value.limit),
        "remaining": nullable(value.remaining),
      ]
    } else {
      usage = NSNull()
    }

    return [
      "success": result.success,
      "featureId": result.featureId,
      "amountUsed": result.amountUsed,
      "message": nullable(result.message),
      "usage": usage,
      "authoritativeAccess": nullable(result.authoritativeAccess.map(featureAccessDictionary)),
    ]
  }
  #endif

  private func blocking<T>(
    _ operation: @escaping @Sendable () async throws -> T
  ) -> Result<T, Error> {
    let semaphore = DispatchSemaphore(value: 0)
    let box = BlockingResultBox<T>()
    Task.detached {
      do {
        box.store(.success(try await operation()))
      } catch {
        box.store(.failure(error))
      }
      semaphore.signal()
    }
    semaphore.wait()
    return box.load() ?? .failure(
      NSError(
        domain: "ai.nuxie.unity",
        code: -1,
        userInfo: [NSLocalizedDescriptionKey: "Native operation did not produce a result."]
      )
    )
  }

  private func mapResult<T>(
    _ result: Result<T, Error>,
    transform: (T) -> Any? = { $0 }
  ) -> String {
    switch result {
    case .success(let value):
      return okResponse(transform(value))
    case .failure(let error):
      return errorResponse(
        code: "NATIVE_ERROR",
        message: error.localizedDescription,
        nativeStack: String(describing: error)
      )
    }
  }

  private func okResponse(_ value: Any?) -> String {
    serialize(["ok": true, "value": nullable(value)])
  }

  private func errorResponse(
    code: String,
    message: String,
    nativeStack: String? = nil
  ) -> String {
    var error: [String: Any] = ["code": code, "message": message]
    if let nativeStack {
      error["nativeStack"] = nativeStack
    }
    return serialize(["ok": false, "error": error])
  }

  private func serialize(_ value: Any) -> String {
    guard JSONSerialization.isValidJSONObject(value),
          let data = try? JSONSerialization.data(withJSONObject: value),
          let string = String(data: data, encoding: .utf8)
    else {
      return "{\"ok\":false,\"error\":{\"code\":\"NATIVE_ERROR\",\"message\":\"JSON serialization failed.\"}}"
    }
    return string
  }

  private func emit(_ type: String, _ payload: [String: Any]) {
    let message = serialize([
      "type": type,
      "timestampMs": Int(Date().timeIntervalSince1970 * 1_000),
      "payload": payload,
    ])
    let (object, method) = callbackTarget()
    DispatchQueue.main.async {
      sendUnityMessage(object: object, method: method, payload: message)
    }
  }
}

private final class UnsafeSendable<Value>: @unchecked Sendable {
  let value: Value

  init(_ value: Value) {
    self.value = value
  }
}

private final class BlockingResultBox<Value>: @unchecked Sendable {
  private let lock = NSLock()
  private var value: Result<Value, Error>?

  func store(_ value: Result<Value, Error>) {
    lock.withLock {
      self.value = value
    }
  }

  func load() -> Result<Value, Error>? {
    lock.withLock { value }
  }
}

private func nullable(_ value: Any?) -> Any {
  value ?? NSNull()
}

#if canImport(Nuxie)
@MainActor
private final class UnityDelegateBridge: NuxieDelegate {
  private let emit: @Sendable (String, [String: Any]) -> Void

  init(emit: @escaping @Sendable (String, [String: Any]) -> Void) {
    self.emit = emit
  }

  func featureAccessDidChange(
    _ featureId: String,
    from oldValue: FeatureAccess?,
    to newValue: FeatureAccess
  ) {
    emit("feature_access_changed", [
      "featureId": featureId,
      "from": nullable(oldValue.map(featureAccessDictionary)),
      "to": featureAccessDictionary(newValue),
    ])
  }

  func nuxieDidEmit(_ info: NuxieActivityInfo) {
    emit("activity", [
      "schemaVersion": NuxieActivityInfo.schemaVersion,
      "id": info.id,
      "timestampMs": Int(info.timestamp.timeIntervalSince1970 * 1_000),
      "receivedAtMs": Int(info.receivedAt.timeIntervalSince1970 * 1_000),
      "name": info.name,
      "properties": info.properties.mapValues(activityValue),
    ])
  }

  func nuxie(_ sdk: NuxieSDK, didRequestAppAction action: AppAction) {
    emit("app_action", [
      "name": action.name,
      "payload": nullable(action.payload?.mapValues(appActionValue)),
      "experience": [
        "experienceId": action.experience.experienceId,
        "experienceVersion": nullable(action.experience.experienceVersion),
        "journeyId": nullable(action.experience.journeyId),
      ],
    ])
  }
}

private func featureAccessDictionary(_ access: FeatureAccess) -> [String: Any] {
  [
    "allowed": access.allowed,
    "unlimited": access.unlimited,
    "balance": nullable(access.balance),
    "type": access.type.rawValue,
  ]
}

private func activityValue(_ value: NuxieActivityValue) -> Any {
  switch value {
  case .string(let value): value
  case .int(let value): value
  case .double(let value): value
  case .bool(let value): value
  }
}

private func appActionValue(_ value: AppActionValue) -> Any {
  switch value {
  case .string(let value): value
  case .int(let value): value
  case .double(let value): value
  case .bool(let value): value
  }
}

private final class UnityPurchaseDelegateBridge: NuxiePurchaseDelegate, @unchecked Sendable {
  private let emit: @Sendable (String, [String: Any]) -> Void
  private let timeoutSeconds: TimeInterval
  private let lock = NSLock()
  private var purchases: [String: CheckedContinuation<PurchaseResult, Never>] = [:]
  private var restores: [String: CheckedContinuation<RestoreResult, Never>] = [:]

  init(
    timeoutSeconds: TimeInterval = 60,
    emit: @escaping @Sendable (String, [String: Any]) -> Void
  ) {
    self.timeoutSeconds = timeoutSeconds
    self.emit = emit
  }

  func purchase(product: StoreProduct) async -> PurchaseResult {
    let requestId = UUID().uuidString
    return await withCheckedContinuation { continuation in
      lock.withLock {
        purchases[requestId] = continuation
      }
      emit("purchase_request", [
        "request_id": requestId,
        "platform": "ios",
        "product_id": product.productId,
        "store_product_id": product.storeProductId,
        "base_plan_id": NSNull(),
        "purchase_option_id": NSNull(),
        "offer_id": NSNull(),
        "placement_id": nullable(product.placementId),
        "display_name": product.name,
        "display_price": product.price,
        "timestamp_ms": Int(Date().timeIntervalSince1970 * 1_000),
      ])
      schedulePurchaseTimeout(requestId)
    }
  }

  func restorePurchases() async -> RestoreResult {
    let requestId = UUID().uuidString
    return await withCheckedContinuation { continuation in
      lock.withLock {
        restores[requestId] = continuation
      }
      emit("restore_request", [
        "request_id": requestId,
        "platform": "ios",
        "timestamp_ms": Int(Date().timeIntervalSince1970 * 1_000),
      ])
      scheduleRestoreTimeout(requestId)
    }
  }

  func completePurchase(requestId: String, payload: [String: Any]) {
    let continuation = lock.withLock { purchases.removeValue(forKey: requestId) }
    continuation?.resume(returning: purchaseResult(payload))
  }

  func completeRestore(requestId: String, payload: [String: Any]) {
    let continuation = lock.withLock { restores.removeValue(forKey: requestId) }
    continuation?.resume(returning: restoreResult(payload))
  }

  func cancelPending(reason: String) {
    let pending = lock.withLock {
      let pending = (Array(purchases.values), Array(restores.values))
      purchases.removeAll()
      restores.removeAll()
      return pending
    }
    pending.0.forEach { $0.resume(returning: .failed(bridgeError(reason))) }
    pending.1.forEach { $0.resume(returning: .failed(bridgeError(reason))) }
  }

  private func schedulePurchaseTimeout(_ requestId: String) {
    Task { [weak self] in
      guard let self else { return }
      try? await Task.sleep(nanoseconds: UInt64(self.timeoutSeconds * 1_000_000_000))
      let continuation = self.lock.withLock {
        self.purchases.removeValue(forKey: requestId)
      }
      continuation?.resume(returning: .failed(self.bridgeError("purchase_timeout")))
    }
  }

  private func scheduleRestoreTimeout(_ requestId: String) {
    Task { [weak self] in
      guard let self else { return }
      try? await Task.sleep(nanoseconds: UInt64(self.timeoutSeconds * 1_000_000_000))
      let continuation = self.lock.withLock {
        self.restores.removeValue(forKey: requestId)
      }
      continuation?.resume(returning: .failed(self.bridgeError("restore_timeout")))
    }
  }

  private func purchaseResult(_ payload: [String: Any]) -> PurchaseResult {
    switch (payload["type"] as? String)?.lowercased() {
    case "purchased": .purchased
    case "cancelled": .cancelled
    case "pending": .pending
    default: .failed(bridgeError((payload["message"] as? String) ?? "purchase_failed"))
    }
  }

  private func restoreResult(_ payload: [String: Any]) -> RestoreResult {
    switch (payload["type"] as? String)?.lowercased() {
    case "restored": .restored
    case "no_purchases": .noPurchases
    default: .failed(bridgeError((payload["message"] as? String) ?? "restore_failed"))
    }
  }

  private func bridgeError(_ message: String) -> Error {
    NSError(
      domain: "ai.nuxie.unity",
      code: 1,
      userInfo: [NSLocalizedDescriptionKey: message]
    )
  }
}
#endif

private func sendUnityMessage(object: String, method: String, payload: String) {
  object.withCString { objectPointer in
    method.withCString { methodPointer in
      payload.withCString { payloadPointer in
        UnitySendMessage(objectPointer, methodPointer, payloadPointer)
      }
    }
  }
}

@_silgen_name("UnitySendMessage")
private func UnitySendMessage(
  _ object: UnsafePointer<CChar>,
  _ method: UnsafePointer<CChar>,
  _ message: UnsafePointer<CChar>
)

File: Runtime/Plugins/Android/nuxie-unity-bridge.androidlib/src/main/kotlin/ai/nuxie/unity/NuxieUnityBridge.kt
package ai.nuxie.unity

import ai.nuxie.sdk.AppAction
import ai.nuxie.sdk.AppActionValue
import ai.nuxie.sdk.LogLevel
import ai.nuxie.sdk.Nuxie
import ai.nuxie.sdk.NuxieActivityInfo
import ai.nuxie.sdk.NuxieActivityValue
import ai.nuxie.sdk.NuxieConfiguration
import ai.nuxie.sdk.NuxieEnvironment
import ai.nuxie.sdk.NuxieListener
import ai.nuxie.sdk.billing.NuxiePurchaseDelegate
import ai.nuxie.sdk.billing.PurchaseHandlingMode
import ai.nuxie.sdk.billing.PurchaseResult
import ai.nuxie.sdk.billing.RestoreResult
import ai.nuxie.sdk.billing.StoreProduct
import ai.nuxie.sdk.features.FeatureAccess
import ai.nuxie.sdk.features.FeatureCheckPolicy
import ai.nuxie.sdk.features.FeatureType
import ai.nuxie.sdk.features.FeatureUsageResult
import com.unity3d.player.UnityPlayer
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withTimeoutOrNull
import org.json.JSONArray
import org.json.JSONObject

object NuxieUnityBridge {
  private val purchaseDelegate = UnityPurchaseDelegate(::emit)
  private val listener = object : NuxieListener {
    override fun featureAccessDidChange(
      featureId: String,
      oldAccess: FeatureAccess?,
      newAccess: FeatureAccess,
    ) {
      emit(
        "feature_access_changed",
        mapOf(
          "featureId" to featureId,
          "from" to oldAccess?.toMap(),
          "to" to newAccess.toMap(),
        ),
      )
    }

    override fun onActivityEmitted(sdk: Nuxie, info: NuxieActivityInfo) {
      emit("activity", info.toMap())
    }

    override fun onAppActionRequested(sdk: Nuxie, action: AppAction) {
      emit("app_action", action.toMap())
    }
  }

  @Volatile
  private var callbackObject = "__NuxieBridgeHost"

  @Volatile
  private var callbackMethod = "OnNuxieNativeEvent"

  @JvmStatic
  fun invoke(
    method: String,
    argumentsJson: String,
    callbackObject: String,
    callbackMethod: String,
  ): String {
    this.callbackObject = callbackObject
    this.callbackMethod = callbackMethod

    return runCatching {
      val arguments = jsonObjectToMap(JSONObject(argumentsJson))
      when (method) {
        "configure" -> configure(arguments)
        "shutdown" -> {
          purchaseDelegate.cancelPending("sdk_shutdown")
          runBlocking { Nuxie.shutdown() }
          if (Nuxie.listener === listener) {
            Nuxie.listener = null
          }
          okResponse(null)
        }

        "identify" -> {
          val distinctId = arguments["distinctId"] as? String
            ?: return@runCatching errorResponse(
              "INVALID_ARGUMENTS",
              "identify requires distinctId.",
            )
          Nuxie.identify(
            distinctId = distinctId,
            userProperties = arguments["userProperties"].asStringAnyMap(),
            userPropertiesSetOnce = arguments["userPropertiesSetOnce"].asStringAnyMap(),
          )
          okResponse(null)
        }

        "reset" -> {
          Nuxie.reset(keepAnonymousId = arguments["keepAnonymousId"] as? Boolean ?: false)
          okResponse(null)
        }

        "getDistinctId" -> okResponse(Nuxie.distinctId)
        "getAnonymousId" -> okResponse(Nuxie.anonymousId)
        "getIsIdentified" -> okResponse(Nuxie.isIdentified)

        "trigger" -> {
          val eventName = arguments["eventName"] as? String
            ?: return@runCatching errorResponse(
              "INVALID_ARGUMENTS",
              "trigger requires eventName.",
            )
          Nuxie.trigger(eventName, arguments["properties"].asStringAnyMap())
          okResponse(null)
        }

        "dismiss" -> {
          runBlocking { Nuxie.dismiss() }
          okResponse(null)
        }

        "setLocaleIdentifier" -> {
          runBlocking {
            Nuxie.setLocaleIdentifier(arguments["localeIdentifier"] as? String)
          }
          okResponse(null)
        }

        "hasFeature" -> {
          val featureId = arguments["featureId"] as? String
            ?: return@runCatching errorResponse(
              "INVALID_ARGUMENTS",
              "hasFeature requires featureId.",
            )
          val access = runBlocking {
            Nuxie.hasFeature(
              featureId = featureId,
              requiredBalance = (arguments["requiredBalance"] as? Number)?.toDouble() ?: 1.0,
              entityId = arguments["entityId"] as? String,
              policy = if (arguments["policy"] == "remote") {
                FeatureCheckPolicy.REMOTE
              } else {
                FeatureCheckPolicy.CACHE_FIRST
              },
            )
          }
          okResponse(access.toMap())
        }

        "useFeature" -> {
          val featureId = arguments["featureId"] as? String
            ?: return@runCatching errorResponse(
              "INVALID_ARGUMENTS",
              "useFeature requires featureId.",
            )
          Nuxie.useFeature(
            featureId = featureId,
            amount = (arguments["amount"] as? Number)?.toDouble() ?: 1.0,
            entityId = arguments["entityId"] as? String,
            metadata = arguments["metadata"].asStringAnyMap(),
          )
          okResponse(null)
        }

        "useFeatureAndWait" -> {
          val featureId = arguments["featureId"] as? String
            ?: return@runCatching errorResponse(
              "INVALID_ARGUMENTS",
              "useFeatureAndWait requires featureId.",
            )
          val result = runBlocking {
            Nuxie.useFeatureAndWait(
              featureId = featureId,
              amount = (arguments["amount"] as? Number)?.toDouble() ?: 1.0,
              entityId = arguments["entityId"] as? String,
              setUsage = arguments["setUsage"] as? Boolean ?: false,
              metadata = arguments["metadata"].asStringAnyMap(),
            )
          }
          okResponse(result.toMap())
        }

        "completePurchase" -> {
          val requestId = arguments["requestId"] as? String
            ?: return@runCatching errorResponse(
              "INVALID_ARGUMENTS",
              "completePurchase requires requestId.",
            )
          val result = arguments["result"].asStringAnyMap()
            ?: return@runCatching errorResponse(
              "INVALID_ARGUMENTS",
              "completePurchase requires result.",
            )
          purchaseDelegate.completePurchase(requestId, result)
          okResponse(null)
        }

        "completeRestore" -> {
          val requestId = arguments["requestId"] as? String
            ?: return@runCatching errorResponse(
              "INVALID_ARGUMENTS",
              "completeRestore requires requestId.",
            )
          val result = arguments["result"].asStringAnyMap()
            ?: return@runCatching errorResponse(
              "INVALID_ARGUMENTS",
              "completeRestore requires result.",
            )
          purchaseDelegate.completeRestore(requestId, result)
          okResponse(null)
        }

        else -> errorResponse("NATIVE_ERROR", "Unsupported method '$method'.")
      }
    }.getOrElse { error ->
      errorResponse(
        "NATIVE_ERROR",
        error.message ?: "Native bridge failure.",
        error.stackTraceToString(),
      )
    }
  }

  private fun configure(arguments: Map<String, Any?>): String {
    val apiKey = arguments["apiKey"] as? String
      ?: return errorResponse("MISSING_API_KEY", "Nuxie API key is required.")
    if (apiKey.isBlank()) {
      return errorResponse("MISSING_API_KEY", "Nuxie API key is required.")
    }

    val context = UnityPlayer.currentActivity?.applicationContext
      ?: return errorResponse("NO_CONTEXT", "Unity activity is unavailable.")
    val options = arguments["options"].asStringAnyMap()
    val usePurchaseController = arguments["usingPurchaseController"] as? Boolean ?: false

    Nuxie.listener = listener
    try {
      Nuxie.setup(
        context,
        NuxieConfiguration(apiKey).apply {
        environment = if (options?.get("environment") == "development") {
          NuxieEnvironment.DEVELOPMENT
        } else {
          NuxieEnvironment.PRODUCTION
        }
        logLevel = when (options?.get("logLevel") as? String) {
          "verbose", "debug" -> LogLevel.DEBUG
          "info" -> LogLevel.INFO
          "error" -> LogLevel.ERROR
          "none" -> LogLevel.NONE
          else -> LogLevel.WARN
        }
        if (options?.containsKey("localeIdentifier") == true) {
          localeIdentifier = options["localeIdentifier"] as? String
        }
        purchaseHandlingMode = if (options?.get("purchaseHandlingMode") == "observer") {
          PurchaseHandlingMode.APP_MANAGED
        } else {
          PurchaseHandlingMode.NUXIE_MANAGED
        }
        if (usePurchaseController) {
          purchaseDelegate = this@NuxieUnityBridge.purchaseDelegate
        }
        },
      )
    } catch (error: Throwable) {
      if (!Nuxie.isSetup && Nuxie.listener === listener) {
        Nuxie.listener = null
      }
      throw error
    }
    return okResponse(null)
  }

  private fun emit(type: String, payload: Map<String, Any?>) {
    val message = JSONObject(
      mapOf(
        "type" to type,
        "timestampMs" to System.currentTimeMillis(),
        "payload" to payload,
      ),
    ).toString()

    val objectName = callbackObject
    val methodName = callbackMethod
    val activity = UnityPlayer.currentActivity
    if (activity == null) {
      UnityPlayer.UnitySendMessage(objectName, methodName, message)
    } else {
      activity.runOnUiThread {
        UnityPlayer.UnitySendMessage(objectName, methodName, message)
      }
    }
  }

  private fun okResponse(value: Any?): String = JSONObject()
    .put("ok", true)
    .put("value", value ?: JSONObject.NULL)
    .toString()

  private fun errorResponse(
    code: String,
    message: String,
    nativeStack: String? = null,
  ): String {
    val error = JSONObject()
      .put("code", code)
      .put("message", message)
    if (nativeStack != null) {
      error.put("nativeStack", nativeStack)
    }
    return JSONObject()
      .put("ok", false)
      .put("error", error)
      .toString()
  }
}

private class UnityPurchaseDelegate(
  private val emit: (String, Map<String, Any?>) -> Unit,
  private val timeoutMs: Long = 60_000,
) : NuxiePurchaseDelegate {
  private val purchases = ConcurrentHashMap<String, CompletableDeferred<PurchaseResult>>()
  private val restores = ConcurrentHashMap<String, CompletableDeferred<RestoreResult>>()

  override suspend fun purchase(product: StoreProduct): PurchaseResult {
    val requestId = UUID.randomUUID().toString()
    val deferred = CompletableDeferred<PurchaseResult>()
    purchases[requestId] = deferred

    emit(
      "purchase_request",
      mapOf(
        "request_id" to requestId,
        "platform" to "android",
        "product_id" to product.productId,
        "store_product_id" to product.storeProductId,
        "base_plan_id" to product.basePlanId,
        "purchase_option_id" to product.purchaseOptionId,
        "offer_id" to product.offerId,
        "placement_id" to product.placementId,
        "display_name" to product.rawProduct?.name,
        "display_price" to null,
        "timestamp_ms" to System.currentTimeMillis(),
      ),
    )

    return try {
      withTimeoutOrNull(timeoutMs) { deferred.await() }
        ?: PurchaseResult.Failed(bridgeError("purchase_timeout"))
    } finally {
      purchases.remove(requestId)
    }
  }

  override suspend fun restorePurchases(): RestoreResult {
    val requestId = UUID.randomUUID().toString()
    val deferred = CompletableDeferred<RestoreResult>()
    restores[requestId] = deferred

    emit(
      "restore_request",
      mapOf(
        "request_id" to requestId,
        "platform" to "android",
        "timestamp_ms" to System.currentTimeMillis(),
      ),
    )

    return try {
      withTimeoutOrNull(timeoutMs) { deferred.await() }
        ?: RestoreResult.Failed(bridgeError("restore_timeout"))
    } finally {
      restores.remove(requestId)
    }
  }

  fun completePurchase(requestId: String, payload: Map<String, Any?>) {
    purchases.remove(requestId)?.complete(
      when ((payload["type"] as? String)?.lowercase()) {
        "purchased" -> PurchaseResult.Purchased
        "cancelled" -> PurchaseResult.Cancelled
        "pending" -> PurchaseResult.Pending
        else -> PurchaseResult.Failed(
          bridgeError((payload["message"] as? String) ?: "purchase_failed"),
        )
      },
    )
  }

  fun completeRestore(requestId: String, payload: Map<String, Any?>) {
    restores.remove(requestId)?.complete(
      when ((payload["type"] as? String)?.lowercase()) {
        "restored" -> RestoreResult.Restored
        "no_purchases" -> RestoreResult.NoPurchases
        else -> RestoreResult.Failed(
          bridgeError((payload["message"] as? String) ?: "restore_failed"),
        )
      },
    )
  }

  fun cancelPending(reason: String) {
    purchases.values.forEach { it.complete(PurchaseResult.Failed(bridgeError(reason))) }
    restores.values.forEach { it.complete(RestoreResult.Failed(bridgeError(reason))) }
    purchases.clear()
    restores.clear()
  }

  private fun bridgeError(message: String): Throwable = IllegalStateException(message)
}

private fun FeatureType.toBridgeValue(): String = when (this) {
  FeatureType.BOOLEAN -> "boolean"
  FeatureType.METERED -> "metered"
  FeatureType.CREDIT_SYSTEM -> "creditSystem"
}

private fun FeatureAccess.toMap(): Map<String, Any?> = mapOf(
  "allowed" to allowed,
  "unlimited" to unlimited,
  "balance" to balance,
  "type" to type.toBridgeValue(),
)

private fun FeatureUsageResult.toMap(): Map<String, Any?> = mapOf(
  "success" to success,
  "featureId" to featureId,
  "amountUsed" to amountUsed,
  "message" to message,
  "usage" to usage?.let { value ->
    mapOf(
      "current" to value.current,
      "limit" to value.limit,
      "remaining" to value.remaining,
    )
  },
  "authoritativeAccess" to authoritativeAccess?.toMap(),
)

private fun NuxieActivityInfo.toMap(): Map<String, Any?> = mapOf(
  "schemaVersion" to NuxieActivityInfo.SCHEMA_VERSION,
  "id" to id,
  "timestampMs" to timestampMillis,
  "receivedAtMs" to receivedAtMillis,
  "name" to name,
  "properties" to properties.mapValues { (_, value) -> value.toBridgeValue() },
)

private fun NuxieActivityValue.toBridgeValue(): Any = when (this) {
  is NuxieActivityValue.String -> value
  is NuxieActivityValue.Int -> value
  is NuxieActivityValue.Double -> value
  is NuxieActivityValue.Bool -> value
}

private fun AppAction.toMap(): Map<String, Any?> = mapOf(
  "name" to name,
  "payload" to payload?.mapValues { (_, value) -> value.toBridgeValue() },
  "experience" to mapOf(
    "experienceId" to experience.experienceId,
    "experienceVersion" to experience.experienceVersion,
    "journeyId" to experience.journeyId,
  ),
)

private fun AppActionValue.toBridgeValue(): Any = when (this) {
  is AppActionValue.String -> value
  is AppActionValue.Int -> value
  is AppActionValue.Double -> value
  is AppActionValue.Bool -> value
}

private fun jsonObjectToMap(value: JSONObject): Map<String, Any?> {
  val result = mutableMapOf<String, Any?>()
  val keys = value.keys()
  while (keys.hasNext()) {
    val key = keys.next()
    result[key] = jsonValue(value.get(key))
  }
  return result
}

private fun jsonArrayToList(value: JSONArray): List<Any?> =
  (0 until value.length()).map { index -> jsonValue(value.get(index)) }

private fun jsonValue(value: Any?): Any? = when (value) {
  null, JSONObject.NULL -> null
  is JSONObject -> jsonObjectToMap(value)
  is JSONArray -> jsonArrayToList(value)
  else -> value
}

@Suppress("UNCHECKED_CAST")
private fun Any?.asStringAnyMap(): Map<String, Any?>? = this as? Map<String, Any?>
